### DISCUSSION TITLE
[Suggestion] Cargo Proposal

### DISCUSSION THREAD

**Discussion by Dwarf-LordPangolin** [tsunamayo/Starship-EVO#4098] [Suggestions]
I think I speak for all the testers in saying that we love the idea of items being carried as physical cargo in ships' holds, rather than being stored in a magic "inventory" for the ship. This might be challenging for the game's AI, so this is a suggestion for one way to make it easy for NPCs to load & unload cargo on ships.

If items are stored as in-game objects, the AI is going to need some method of knowing where a ship's cargo hold is. This either requires a complex algorithm that might still get it wrong (such as storing cargo in a ship's hangar bay by mistake), or -- preferably -- a performance-friendly way of telling the AI "put cargo here." Crates and Shelves are one method of doing this.

![20190621034821_1](https://user-images.githubusercontent.com/33041786/59940018-2fe5d800-941f-11e9-83f8-51ed1e91d023.jpg)
_A Crate_

Items could be put inside Crates by activating the Crate and transferring items from the player's inventory to the Crate's. Crates could be placed as Blocks, or carried by players and NPCs one at a time in their arms. A player or NPC could activate a Crate to either open it or carry it. Crates would visually be boxes about 1 cubic meter (slightly less to prevent clipping), and could carry 1 cubic meter's worth of items. Crates could only be stacked vertically on Shelves.

![20190621033713_1](https://user-images.githubusercontent.com/33041786/59940109-75a2a080-941f-11e9-9f67-7de4e3be6703.jpg)
_Shelves with Crates_

Shelves would be blocks designated for the AI as areas to place Crates. Shelves would look a little like the Grating_A block. A player or NPC carrying a Crate could activate a Shelf to place that Crate on that Shelf. Activating a Shelf while not carrying a Crate would let the player or NPC chose which Crate on that Shelf to pick up. Activating the Shelf rather than the Crate means the desired Crate can be gotten to without requiring everything between it and the player/NPC to be unloaded first, which would be challenging for the game's AI and annoying for the player. All Shelf blocks adjacent to each other would function as a unit; activating one would allow the player/NPC to place or pick up any Crate on any connected Shelf.

![20190621034110_1](https://user-images.githubusercontent.com/33041786/59940296-f2357f00-941f-11e9-8bfe-1e18bb3caaae.jpg)
_Activating the Shelf means the poor AI does not have to navigate a potential maze of shelving, possibly getting blocked or stuck_

This system makes it very easy for the AI to handle loading & unloading of cargo. At a port, the player selects an NPC that works for them, give them a "Load this" command, select a Crate or Crates, then select which of their owned ships at that port they want it loaded onto. The AI then checks that ship for any Shelves, finds a path to them for the NPC, and tells it to carry that Crate to one of those Shelves. Unloading is similarly easy. The player selects an NPC, gives it a "Unload this" command, selects a Crate or Crates, then selects the destination (either the port itself, or another ship). The AI then does the same check for Shelves, pathfinding check, and then gives the NPC the command.

Using Shelves also means that Crates with other Crates on top of them can be picked up without leaving silly-looking floating Crates; it would simply look like the "floating" Crate was held up by the Shelf.

![20190621035335_1](https://user-images.githubusercontent.com/33041786/59940343-11cca780-9420-11e9-8804-e8ea2e433c28.jpg)

If needed, Large Crates & Large Shelves can also be implemented to reduce the number of containers the game needs to keep track of. Being larger than the player or NPCs, these could be carried by vehicles with Large Shelves; activating a Large Shelf while driving a vehicle with empty Large Shelves would give the player/NPC the option of transferring Large Crates to their vehicle's Large Shelves.

![20190621124806_1](https://user-images.githubusercontent.com/33041786/59941455-d5e71180-9422-11e9-9f95-586378b936ab.jpg)
_Not only is this efficient storage, it also gives us an excuse to build big yellow loader mechs!_

### Conclusion
Crates would give the game a portable way of storing cargo as a physical and visible in-game object. Shelves would give NPCs a designated area to place cargo, allowing the game's AI to handle loading and unloading of cargo simply by using its pathfinding ability to reach an area, and then use an interaction on an in-game object.

**Comment by ZachZent** [comment]
Do have a few additions to this excellent idea which would make it a bit more expansive without it being too complicated. First is the basic materials.

I think it can be safely assumed that there will be solid, liquid, and gas materials. Perhaps even more... exotic materials like black hole juice (a technical term.) Rather then having one chest that can hold everything, say a bottle of super heated plasma next to space weed, there would be a chest for each "material type." A normal chest for physical items like blocks or cargo, a gas container for gas extracted or bought, a liquid container for water or fuel, and finally an optional "exotic" chest for the special materials that probably have the the word quantum in the name. I don't think it should go much more complicated past that. Like a refrigeration chest for foods stuffs. At least not right away or to leave it for mods. This more in depth yet comprehensible system would allow for a bit more personalized management and a need for planning on how your ship will transport and store what cargo. For example ship that mine gas giants would have to carry more gas canisters while a trade ship would carry more standard chests.

""and could carry 1 cubic meter's worth of items.""
Perhaps I am misunderstanding what Dwarf is saying about this, but if a chest can only hold basically the chest's sizes worth means that carrying lots of stuff is impossible without a 500m cargo ship to carry basically a single "stack" of blocks. Small cargo ships need love too. It may not be as realistic, but is is much more reasonable. If this is not what Dwarf is saying, then ignore above but the idea below stands. As such, second is the chest's capacity.

Chests should act like "enhanced" Minecraft chests. Each single chest has a number of slots. Unlike Minecraft where the stacks are always 64, stack sizes will depend on the mass of the blocks. Say standard hull stacks up to 100, engines to 25, and so on. The larger mech-sized chests have more slots and whatnot. The other chest types are based on volume/density. Perhaps the exotic chest can only hold one thing. I don't think dark matter should exactly be put right next to element 547...

The idea for the carrying capacity of the chests are similar to Minecraft. There are two possibilities to add this style of chests to Skywanderes. The first version is that chests are equivalent to the standard Minecraft chests. Chests are of only one type, of one carrying capacity. No matter what you craft it from, spruce or oak wood, there are always 24 slots open. The second version of chests resemble Minecraft modded chests, specifically the iron chest mod (http://www.9minecraft.net/iron-chests-mod/) While the vanilla chests are of one size alone, the modded chests have multiple carrying capacities. Though more and more expensive materials in crafting the chest, the capacity grows all while the size of the chests stay the same of 1x1x1m. Either the number of slots increase or the amount of mass per slot increases. In the mod there are many chests types, but there doesn't need to be that many in game. Say three versions of the normal 1x1x1m chest and three versions of the larger mech-sized chests with increasing capacity for both. Personally I would like the more expansive version for more varied cargo, but that is up to you (Tsuna.)

The main shelf idea is great and fixes a lot of the possible issues that could arise from adding physical (well chest-physical) cargo. This is just an expansion to the idea that I hope you consider or use as inspiration.

(Edited for clarity)

**Comment by tsunamayo** [comment]
Very interesting ideas, I keep that in mind for when I come to this. The biggest hurdle before I can get to that is more on the npc side tbh. Yes having physical inventory looks more sexy for sure. I had in mind something physical also that you can carry, but maybe requiring an exo suit of some sort, but unfortunately I am far far from this. At first I want to make a combat focused experience because that is what people expect and want to build, but then I will want to expand and add depth to do interesting cargo hauling gameplay (like with maybe some smuggling ect).
For the storage issue mentionned by Zach, this is something I had in mind: I wanted to add various tier of storage crate to of course stack more than one block in the crate. But the important part is that those crate would consume energy, (lets say to condense matter from the lore perspective). So that you have to decide how to spend your energy on the ship, a ship with more cargo capacity for mining or scanvenging will be less powerfull in combat, ect. That would give rise to dedicated cargo ship, and interesting trade-off choice when designing the ship.
Thanks!
